### PR TITLE
Users/prmutn/bug/668516

### PR DESCRIPTION
**Pull request by preetsaimutneja**
fix for #668537 (VSTest parallel execution feature is not working with willow in node implementation)  and #668516 (VSTest Task doesn't recognize DLL in source path with filter).

### CODE REVIEW

File: Tasks/VsTest/vstest.ts [hunk 1]
```diff
@@ -1211,19 +1226,28 @@ function saveToFile(fileContents: string, extension: string): Q.Promise<string>
...
+        if (vstestLocationMethod.toLowerCase() === 'version') {
+            if (!isNaN(vsVersion) && vsVersion >= 14) {
+                if (vsVersion >= 15) { // moved away from taef
+                    return;
+                }
 
-            // in 14.0 taef parellization needed taef enabled
-            var vs14Common = tl.getVariable("VS140COMNTools");
+                // in 14.0 taef parellization needed taef enabled
+                let vs14Common: string = tl.getVariable("VS140COMNTools");
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: what happen if i have only VS15 willow installed? will this work?

File: Tasks/VsTest/task.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "version": {
         "Major": 1,
         "Minor": 0,
-        "Patch": 66
+        "Patch": 68
```
Comment (nigurr) [inline Tasks/VsTest/task.json]: why 68?

File: Tasks/VsTest/vstest.ts [hunk 3]
```diff
@@ -84,34 +84,49 @@ catch (error) {
...
+    }
+}
+
+function getPatternWithoutIncludeExclude(pattern: string): string {
+    return pattern.startsWith('-:') || pattern.startsWith('+:') ? pattern.substr(2) : pattern;
+}
+
 function getTestAssemblies(): Set<string> {
-    var testAssemblyFiles = [];
+    let testAssemblyFiles: Array<string> = [];
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: it can be simple `string[]`

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -84,34 +84,49 @@ catch (error) {
...
-                if (testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0) {
-                    excludeTestAssemblies = excludeTestAssemblies.concat(getFilteredFiles(testAssemblyFilter.substr(2), allFiles));
-                }
-                else {
-                    excludeTestAssemblies.push(testAssemblyFilter.substr(2));
-                }
-            }
-            else if (testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0) {
-                testAssemblyFiles = testAssemblyFiles.concat(getFilteredFiles(testAssemblyFilter, allFiles));
+            let exclude: boolean = testAssemblyFilter.startsWith('-:');
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: better naming. boolean params recommended to contain `isExcludeFilter` something like this

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -84,34 +84,49 @@ catch (error) {
...
+            let hasQuantifier: boolean = Math.max(testAssembly.indexOf('*'), testAssembly.indexOf('?')) != -1;
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: do we really need `Math` functionality for this?

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -84,34 +84,49 @@ catch (error) {
...
+            let patternWithoutIncludeExclude: string = getPatternWithoutIncludeExclude(testAssemblyFilter);
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: ts will auto-resolve to `string` if it's strongly typed call. not required to `strong` name the variable everytime

File: Tasks/VsTest/vstest.ts [hunk 7]
```diff
@@ -84,34 +84,49 @@ catch (error) {
...
+            if (exclude) {
+                excludeTestAssemblies.push.apply(excludeTestAssemblies, hasQuantifier ? 
+                    getFilteredFiles(resolvedPattern,allFiles) : 
+                    [resolvedPattern]
+                );
             }
             else {
-                testAssemblyFiles.push(testAssemblyFilter);
+                testAssemblyFiles.push.apply(testAssemblyFiles, hasQuantifier ? 
+                    getFilteredFiles(resolvedPattern,allFiles) :
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: same code twice. refactor it

File: Tests/L0/VsTest/_suite.ts [hunk 8]
```diff
@@ -169,7 +169,7 @@ describe('VsTest Suite', function () {
     it('Vstest task with test results files filter and exclude filter', (done) => {
         setResponseFile('vstestGood.json');
         var tr = new trm.TaskRunner('VSTest');
-        tr.setInput('testAssembly', '/some/*pattern;-:/exclude/*pattern');
+        tr.setInput('testAssembly', 'some/*pattern;-:exclude/*pattern');
```
Comment (nigurr) [inline Tests/L0/VsTest/_suite.ts]: what happens if I give filter like this `-:abc\t*dll`, `-:\abc\t*dll`, `-:abc\\t*dll` in ps version? this should be in sync with `ps` version

File: Tasks/VsTest/vstest.ts (hunk 1 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: yes

**Comment by nigurr** [comment]
LGTM

File: Tasks/VsTest/vstest.ts (hunk 3 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: done

File: Tasks/VsTest/vstest.ts (hunk 4 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: done

File: Tasks/VsTest/vstest.ts (hunk 5 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: changed

File: Tasks/VsTest/vstest.ts (hunk 6 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: done

File: Tasks/VsTest/vstest.ts (hunk 7 above)
Comment (preetsaimutneja) [inline Tasks/VsTest/vstest.ts]: done

File: Tasks/VsTest/vstest.ts [hunk 9]
```diff
@@ -1220,19 +1233,28 @@ function saveToFile(fileContents: string, extension: string): Q.Promise<string>
...
+                if (vs14Common && pathExistsAsFile(path.join(vs14Common, "..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\TE.TestModes.dll"))) {
+                    setRegistryKeyForParallelExecution(vsVersion);
+                    return;
+                }
+            }
+            resetRunInParallel();
+        } 
+        else if (vstestLocationMethod.toLowerCase() === 'location') {
+            let vs14Common: string = tl.getVariable("VS140COMNTools");
             if (vs14Common && pathExistsAsFile(path.join(vs14Common, "..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\TE.TestModes.dll"))) {
```
Comment (vimegh) [inline Tasks/VsTest/vstest.ts]: >..\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\TE.TestModes.dll [](start = 70, length = 66)

Will this be agnostic to platform. Else use path seperator

**Comment by vimegh** [comment]
:shipit:

**Pull request merged**